skimage/feature/corner_cy.pyx

```diff
@@ -5,13 +5,258 @@
+
+    Follows the expressions given for hermitian symmetric 3 x 3 matrices in
+    [1]_, but simplified to handle real-valued matrices only.
+
+    Parameters
+    ----------
+    a, d, f, b, e, c : cp.ndarray
+        Images corresponding to the individual components of the matrix, `M`,
+        shown above. For example, ``d = M[0, 1]``.
+    sort : {"ascending", "descending"}, optional
+        Eigenvalues should be sorted in the specified order.
+    abs_sort : boolean, optional
+        If ``True``, sort based on the absolute values.
+
+    References
+    ----------
+    .. [1] C. Deledalle, L. Denis, S. Tabti, F. Tupin. Closed-form expressions
+        of the eigen decomposition of 2 x 2 and 3 x 3 Hermitian matrices.
+        [Research Report] Université de Lyon. 2017.
+        https://hal.archives-ouvertes.fr/hal-01501221/file/matrix_exp_and_log_formula.pdf
+    """  # noqa
+    cdef:
+        # use double-precision intermediate variables for accuracy
+        cnp.float64_t a, b, c, d, e, f, d_sq, e_sq, f_sq;
+        cnp.float64_t x1, x2, phi, tmpa, tmpb, tmpc, arg, x1_term, abc;
+        cnp.float64_t lam1, lam2, lam3, stmp;
+        cnp.float64_t abs_lam1, abs_lam2, abs_lam3;
+        Py_ssize_t planes = M00.shape[0];
+        Py_ssize_t rows = M00.shape[1];
+        Py_ssize_t cols = M00.shape[2];
+        Py_ssize_t pln, row, col;
+
+    if np_floats is cnp.float32_t:
+        dtype = np.float32
+    else:
+        dtype = np.float64
+
+    cdef np_floats[:, :, :, ::1] evals = np.empty((3, planes, rows, cols),
+                                                  dtype=dtype)
+
+    with nogil:
+        for pln in range(planes):
+            for row in range(rows):
+                for col in range(cols):
+                    # M = [[a, d, f],
+                    #      [d, b, e],
+                    #      [f, e, c]]
+                    # so d = M01, etc.
+                    a = <cnp.float64_t>(M00[pln, row, col])
+                    b = <cnp.float64_t>(M11[pln, row, col])
+                    c = <cnp.float64_t>(M22[pln, row, col])
+                    d = <cnp.float64_t>(M01[pln, row, col])
+                    e = <cnp.float64_t>(M12[pln, row, col])
+                    f = <cnp.float64_t>(M02[pln, row, col])
+                    d_sq = d * d
+                    e_sq = e * e
+                    f_sq = f * f
+                    tmpa = 2 * a - b - c
+                    tmpb = 2 * b - a - c
+                    tmpc = 2 * c - a - b
+                    x2 = - tmpa * tmpb * tmpc
+                    x2 += 9 * (tmpc * d_sq + tmpb * f_sq + tmpa * e_sq)
+                    x2 -= 54 * d * e * f
+                    x1 = a*a + b*b + c*c - a*b - a*c - b*c + 3 * (d_sq + e_sq + f_sq)
+                    x1 = fmax(x1, 0.0)
+
+                    if x2 == 0.0:
+                        phi = M_PI / 2.0
+                    else:
+                        # added max() here for numerical stability
```

Comment:
Out of curiosity, how do you justify this step?

Reply:
The sqrt of a negative value would give an imaginary-valued eigenvalue, but that is inconsistent with the property that [real, symmetric matrices only have real-valued eigenvalues](https://math.stackexchange.com/questions/354115/prove-that-the-eigenvalues-of-a-real-symmetric-matrix-are-real).

(technically, the real-valued `sqrt` function used here will give NaN in this case rather than an imaginary value)